### PR TITLE
Ability to search entire room history after upgrading room

### LINKED ISSUE #4410: On room upgrade ensure search covers old and new room.
Related to https://github.com/vector-im/riot-web/issues/7166

Currently, once a room has been upgraded, any attempts to search the room are limited to the new room. Search should cover both the new (current) room and any older version rooms.

### PR DESCRIPTION
**Pull request by anoadragon453**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

When searching a subset of rooms, we check each one for any predecessor rooms, and those for any predecessor rooms and so on, and return all results to the client. No change when searching 'all rooms' as predecessor rooms are already included in this (though it shows for "My Room", "My Room" and "My Room". Maybe it should say "My Room - v1", "My Room - v2" and "My Room - v3" instead).

Closes #4410

### CODE REVIEW

File: synapse/api/filtering.py [hunk 1]
```diff
@@ -444,6 +444,9 @@ def lazy_load_members(self):
     def include_redundant_members(self):
         return self.filter_json.get("include_redundant_members", False)
 
+    def add_room_ids(self, room_ids):
```
Comment (richvdh) [inline synapse/api/filtering.py]: Filters are currently immutable, and I think that's a good thing for being able to reason about them. One way to do this would be to have a `with_room_ids` method which returns a *new* Filter, having replaced the room list.

Needs a docstring too, please.

File: synapse/handlers/search.py [hunk 2]
```diff
@@ -139,6 +187,27 @@ def search(self, user, content, batch=None):
 
         room_ids = search_filter.filter_rooms(room_ids)
```
Comment (richvdh) [inline synapse/handlers/search.py]: should this not happen after we add the historical rooms to the filter?

File: synapse/handlers/search.py [hunk 3]
```diff
@@ -139,6 +187,27 @@ def search(self, user, content, batch=None):
...
 
+        # If doing a subset of all rooms seearch, check if any of the rooms
+        # are from an upgraded room, and search their contents as well
+        # XXX: There is the possibility that we don't have a create event for
```
Comment (richvdh) [inline synapse/handlers/search.py]: we ought to have the create event for any room that the user has ever been a member of, so I don't think this is relevant

File: synapse/handlers/search.py [hunk 4]
```diff
@@ -37,6 +37,54 @@ class SearchHandler(BaseHandler):
...
+            room_id (str): The ID of the room to search through.
+
+        Returns:
+            dict of past room IDs as strings
+        """
+
+        historical_room_ids = []
+
+        while True:
+            state_ids = yield self.store.get_current_state_ids(room_id)
```
Comment (richvdh) [inline synapse/handlers/search.py]: I'd suggest adding a `get_room_predecessor` to `StateGroupWorkerStore` alongside `get_room_version`.

File: synapse/handlers/search.py [hunk 5]
```diff
@@ -37,6 +37,54 @@ class SearchHandler(BaseHandler):
...
+        list and then check that room for a m.room.create event and so on
+        until we can no longer find any more previous rooms.
+
+        The full list of all found rooms in then returned.
+
+        Args:
+            room_id (str): The ID of the room to search through.
+
+        Returns:
+            dict of past room IDs as strings
```
Comment (richvdh) [inline synapse/handlers/search.py]: I don't think it's a dict...

Itym `Deferred[list[str]]: predecessor room ids`, or better yet, `Deferred[iterable[str]]: predecessor room ids`

File: synapse/storage/state.py [hunk 6]
```diff
@@ -448,6 +448,7 @@ def get_current_state_ids(self, room_id):
         Returns:
             deferred: dict of (type, state_key) -> event_id
         """
+
```
Comment (richvdh) [inline synapse/storage/state.py]: please try to avoid unrelated reformatting in your PRs; it makes git history harder to read.

File: synapse/storage/state.py [hunk 7]
```diff
@@ -437,6 +437,33 @@ def get_room_version(self, room_id):
...
+            return None
+
+        # Retrieve the room's create event
+        create_event = yield self.get_event(create_id)
+
+        if not create_event:
+            return None
+
+        # Return predecessor if present
+        return create_event.content.get("predecessor", None)
```
Comment (hawkowl) [inline synapse/storage/state.py]: @anoadragon453 you can't return in an inlinecallbacks

Comment (anoadragon453) [reply synapse/storage/state.py]: Ah, you're right, thank you. Still getting used to this execution model.

File: synapse/api/filtering.py [hunk 8]
```diff
@@ -444,6 +444,20 @@ def lazy_load_members(self):
     def include_redundant_members(self):
         return self.filter_json.get("include_redundant_members", False)
 
+    def with_room_ids(self, room_ids):
+        """Returns a new filter with the given room IDs appended.
+
+        Args:
+            room_ids (list): A list of room_ids.
```
Comment (richvdh) [inline synapse/api/filtering.py]: it will work with any iterable, and please specify the types of the entries:
```suggestion
            room_ids (iterable[unicode]): The room_ids to add
```

File: synapse/storage/state.py [hunk 9]
```diff
@@ -437,6 +437,33 @@ def get_room_version(self, room_id):
...
+    @defer.inlineCallbacks
+    def get_room_predecessor(self, room_id):
+        """Get the predecessor room of an upgraded room if one exists.
+        Otherwise return None.
+
+        Args:
+            room_id (str)
+
+        Returns:
+            Deferred[str]: predecessor room id
```
Comment (richvdh) [inline synapse/storage/state.py]:
```suggestion
            Deferred[unicode|None]: predecessor room id
```

File: synapse/handlers/search.py [hunk 10]
```diff
@@ -37,6 +37,41 @@ class SearchHandler(BaseHandler):
...
+        list and then check that room for a m.room.create event and so on
+        until we can no longer find any more previous rooms.
+
+        The full list of all found rooms in then returned.
+
+        Args:
+            room_id (str): id of the room to search through.
+
+        Returns:
+            Deferred[iterable[str]]: predecessor room ids
```
Comment (richvdh) [inline synapse/handlers/search.py]: technically these will be `unicode`s on pythons where they are different things to strs:

```suggestion
            Deferred[iterable[unicode]]: predecessor room ids
```

File: synapse/storage/state.py [hunk 11]
```diff
@@ -437,6 +437,33 @@ def get_room_version(self, room_id):
...
+        create_id = state_ids.get((EventTypes.Create, ""))
+
+        # If we can't find the create event, assume we've hit a dead end
+        if not create_id:
+            defer.returnValue(None)
+
+        # Retrieve the room's create event
+        create_event = yield self.get_event(create_id)
+
+        if not create_event:
```
Comment (richvdh) [inline synapse/storage/state.py]: this is redundant fwiw: get_event cannot return None.
